File: host.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <netinet/ether.h>
#include <pthread.h>
#include <arpa/inet.h>
#include <argp.h>               /* horribly non-standard */

static struct argp_option options[] = {
    {"silent", 's', 0, 0, "receive only, no transmit"},
    {0, 0}
};

struct ether_addr l_ether, r_ether;
int wirenum;
int silent;

static error_t parse_opt(int key, char *arg, struct argp_state *state)
{
    static int argnum = 0;
    
    switch (key) {
    case 's':
        silent = 1;
        break;
    case ARGP_KEY_ARG:
        if (argnum == 0)
            ether_aton_r(arg, &l_ether);
        else if (argnum == 1)
            wirenum = atoi(arg);
        else if (argnum == 2)
            ether_aton_r(arg, &r_ether);
        else
            argp_usage(state);
        argnum++;
        break;
    case ARGP_KEY_END:
        if (argnum != 3)
            argp_usage(state);
        break;
    }
    return 0;
}
static char arg_doc[] = "LL:LL:LL:LL:LL:LL <wirenum> RR:RR:RR:RR:RR:RR";
static char doc_doc[] = "Host - send packets across the 'network'";

static struct argp argp = {options, parse_opt, arg_doc, doc_doc};

void *receive(void *arg)
{
    int i, len, sock = (int)arg;
    unsigned char buf[1500];    /* unsigned so printf %02x works... */
    
    while ((len = recv(sock, buf, sizeof(buf), 0)) > 0) {
        char tmp1[32], tmp2[32];
        struct ether_addr *dst = (struct ether_addr*)buf; /* very crude unpacking */
        struct ether_addr *src = (struct ether_addr*)&buf[6];
        printf("received dgram from %s to %s:", ether_ntoa_r(src, tmp1), ether_ntoa_r(dst, tmp2));
        for (i = 0; i < len; i++)
            printf("%02x ", buf[i]);
        printf("\n");
    }
    printf("lost connection\n");
    exit(0);
}

/* all bytes in AF_UNIX abstact socket names are significant, even 0s,
 * so we need this to be compatible with the python code in 'wires'
 */
int addrlen(struct sockaddr_un *a)
{
    return sizeof(a->sun_family) + 1 + strlen(&a->sun_path[1]);
}
        
int main(int argc, char **argv)
{

    argp_parse(&argp, argc, argv, ARGP_IN_ORDER, 0, 0);

    int s = socket(AF_UNIX, SOCK_SEQPACKET, 0);

    struct sockaddr_un l_unix = {.sun_family = AF_UNIX, .sun_path = {0,}};
    sprintf(&l_unix.sun_path[1], "%s.host-%s (wire %d)", getenv("USER"), argv[1], wirenum);
    bind(s, (struct sockaddr*)&l_unix, addrlen(&l_unix));

    struct sockaddr_un w_unix = {.sun_family = AF_UNIX, .sun_path = {0,}};
    sprintf(&w_unix.sun_path[1], "%s.wire.%d", getenv("USER"), wirenum);
    if (connect(s, (struct sockaddr*)&w_unix, addrlen(&w_unix)) < 0)
        perror("can't connect");

    pthread_t t;
    pthread_create(&t, NULL, receive, (void*)s);

    char packet[60] = {0,};
    memcpy(&packet[0], &r_ether, sizeof(r_ether));
    memcpy(&packet[6], &l_ether, sizeof(l_ether));
    short etype = htons(0x900);
    memcpy(&packet[12], &etype, sizeof(etype));
    strcpy(&packet[14], "\xDE\xAD\xBE\xEF\xDE\xAD\xBE\xEF");

    for (;;) {
        sleep(3);
        if (send(s, packet, sizeof(packet), 0) < sizeof(packet)) {
            printf("xmit: connection lost\n");
            exit(1);
        }
    }
}
    

File: host.py
#!/usr/bin/python
#
# simple program to connect to a "wire" and send ethernet packets between two hosts
#

import sys
import socket
import time
import threading
import string
import getpass
import argparse
import struct

# usage: host <mac> <wire> <remote-mac>

parser = argparse.ArgumentParser(description='Host - send packets across the "network"')
parser.add_argument('mymac', metavar='ll:ll:ll:ll:ll:ll', type=str, nargs=1,
                    help='local MAC address')
parser.add_argument('wire', metavar='W', type=int, nargs=1,
                    help='wire # to connect to')
parser.add_argument('remote', metavar='rr:rr:rr:rr:rr:rr', type=str, nargs=1,
                    help='destination MAC address')
parser.add_argument('--silent', action='store_true', help='receive only, no transmit')
args = parser.parse_args()

def ether_aton(a):
    a = a.replace('-', ':')
    b = map(lambda x: int(x,16), a.split(':'))
    return reduce(lambda x,y: x+y, map(lambda x: struct.pack('B', x), b))

def ether_ntoa(n):
    return string.join(map(lambda x: "%02x" % x, 
                           struct.unpack('6B', n)), ':')

def receive(s):
    while True:
        dgram = s.recv(1500)
        if not dgram:
            print 'lost connection'
            sys.exit(1)
        dst,src = struct.unpack('6s 6s', dgram[0:12])
        print 'received dgram from %s to %s:' % (ether_ntoa(src), ether_ntoa(dst))
        print string.join(map(lambda x: '%02x' % ord(x), buffer(dgram)[:]), ' ')
        print ''

if __name__ == '__main__':
    mymac = ether_aton(args.mymac[0])
    wirenum = args.wire[0]
    remote = ether_aton(args.remote[0])

    s = socket.socket(socket.AF_UNIX, socket.SOCK_SEQPACKET)
    s.bind('\0%s.host-%s (wire %d)' % (getpass.getuser(), ether_ntoa(mymac), wirenum))
    if s.connect_ex('\0%s.wire.%d' % (getpass.getuser(), wirenum)):
        print 'connection error'
        sys.exit(1)


    t = threading.Thread(target=receive, args=[s])
    t.daemon = True                   # so ^C works
    t.start()

    pkt_data = string.join(map(chr, (0xDE,0xAD,0xBE,0xEF,0xDE,0xAD,0xBE,0xEF)), '')
    bogus_ethertype = 0x900
    pkt = struct.pack('!6s 6s H', remote, mymac, bogus_ethertype) + pkt_data
    if len(pkt) < 60:
        pkt = pkt + '\0' * (60-len(pkt))
        
    while True:
        time.sleep(3)
        if not args.silent:
            s.send(pkt)
